pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go

```diff
@@ -0,0 +1,1339 @@
+// information that is not in the raft package -- the current connected state
+// of the store, that we will use the RaftTransport to answer.
+//
+// NB: group membership and connectivity information is communicated to the
+// RangeController via a separate channel, as in the v1 implementation, since
+// it depends on information inside and outside Raft. The methods in Raft are
+// the ones RangeController will call.
+//
+// The implementation should not need to acquire Replica.mu since we have not
+// said anything about the relative lock ordering of Replica.mu and the
+// internal mutexes in RangeController.
+type RaftInterface interface {
+	// FollowerState returns the current state of a follower. The value of
+	// nextUpperBound is populated iff in StateReplicate. All entries >=
+	// nextUpperBound have not yet had MsgApps constructed.
+	//
+	// When a follower transitions from {StateProbe,StateSnapshot} =>
+	// StateReplicate, or was already in StateReplicate but was disconnected and
+	// has now reconnected, we can now start trying to send MsgApps. We should
+	// notice such transitions both in HandleRaftEvent and SetReplicas. We
+	// should *not* construct MsgApps for a StateReplicate follower that is
+	// disconnected -- there is no timeliness guarantee on how long a follower
+	// will stay in StateReplicate despite it being down, and by sending such a
+	// follower MsgApps that are not being received we are defeating flow
+	// control (since we will have advanced nextUpperBound).
+	//
+	// RACv1 also cared about three other cases where the follower behaved as if
+	// it were disconnected (a) paused follower, (b) follower is behind, (c)
+	// follower is inactive (see
+	// replicaFlowControlIntegrationImpl.notActivelyReplicatingTo). (b) and (c)
+	// were needed since it paced at rate of slowest replica, while for regular
+	// work we will in v2 pace at slowest in quorum (and we don't care about
+	// elastic experiencing a hiccup, given it paces at rate of slowest). For
+	// (a), we plan to remove follower pausing. So the v2 code will be
+	// simplified.
+	FollowerState(replicaID roachpb.ReplicaID) (state tracker.StateType, nextUpperBound uint64)
+	FollowerTransportConnected(storeID roachpb.StoreID) bool
+	// HighestEntryIndex is the highest index assigned in the log, and produced
+	// in Ready.Entries(). If there have been no entries produced, since this
+	// replica became the leader, this is the commit index.
+	HighestEntryIndex() uint64
+	// MakeMsgApp is used to construct a MsgApp for entries in [start, end).
+	// REQUIRES: start == nextUpperBound and replicaID is in StateReplicate.
+	// REQUIRES: maxSize > 0.
+	//
+	// If the sum of all entries in [start,end) are <= maxSize, all will be
+	// returned. Else, entries will be returned until, and including, the first
+	// entry that causes maxSize to be equaled or exceeded. This implies at
+	// least one entry will be returned in the MsgApp on success.
```

Comment:
Today, various raft log internals return a prefix exceeding the `maxSize` only if it's a single entry; otherwise it's always `<= maxSize`, and the offender entry gets dropped from the response.

On a few occasions, I considered this logic unfortunate (we waste a read) and corner-casey, and contemplated to simplify it into the one you described: allow exceeding the size always.

Is this nuance critical here, or is an implementation detail?